### PR TITLE
load image for existing texture

### PR DESCRIPTION
**Pull request by jareguo**
for https://github.com/cocos-creator/fireball/issues/6082
Add two new methods for loading image for deserialized textures

### CODE REVIEW

File: cocos/renderer/CCTextureCache.cpp [hunk 1]
```diff
@@ -83,16 +83,43 @@ std::string TextureCache::getDescription() const
...
 public:
-    AsyncStruct(const std::string& fn, std::function<void(Texture2D*)> f) : filename(fn), callback(f), pixelFormat(Texture2D::getDefaultAlphaPixelFormat()), loadSuccess(false) {}
+    AsyncStruct(const std::string& fn, std::function<void(Texture2D*)> f) : filename(fn), callback(f), pixelFormat(Texture2D::getDefaultAlphaPixelFormat()), texture(nullptr), loadSuccess(false) {}
 
     std::string filename;
     std::function<void(Texture2D*)> callback;
     Image image;
     Image imageAlpha;
     Texture2D::PixelFormat pixelFormat;
+    Texture2D* texture; // optional, existing texture to init
```
Comment (jareguo) [inline cocos/renderer/CCTextureCache.cpp]: add a new field on TextureCache::AsyncStruct to indicate which texture should be inited

File: cocos/renderer/CCTextureCache.cpp [hunk 2]
```diff
@@ -141,31 +168,30 @@ void TextureCache::addImageAsync(const std::string &path, const std::function<vo
...
-        Director::getInstance()->getScheduler()->schedule(CC_SCHEDULE_SELECTOR(TextureCache::addImageAsyncCallBack), this, 0, false);
+        if (callback) callback(nullptr);
+        return;
     }
 
-    ++_asyncRefCount;
-
     // generate async struct
     AsyncStruct *data = new (std::nothrow) AsyncStruct(fullpath, callback);
+    data->texture = texture;
```
Comment (jareguo) [inline cocos/renderer/CCTextureCache.cpp]: very much the same as TextureCache::addImageAsync, but will not change TextureCache._textures

File: cocos/renderer/CCTextureCache.cpp [hunk 3]
```diff
@@ -259,38 +285,64 @@ void TextureCache::addImageAsyncCallBack(float dt)
...
-
                 texture->initWithImage(image, asyncStruct->pixelFormat);
                 //parse 9-patch info
                 this->parseNinePatchImage(image, texture, asyncStruct->filename);
 #if CC_ENABLE_CACHE_TEXTURE_DATA
                 // cache the texture file name
                 VolatileTextureMgr::addImageTexture(texture, asyncStruct->filename);
 #endif
-                // cache the texture. retain it, since it is added in the map
-                _textures.insert( std::make_pair(asyncStruct->filename, texture) );
```
Comment (jareguo) [inline cocos/renderer/CCTextureCache.cpp]: dont use texture cache if texture is already specified in AsyncStruct

File: cocos/scripting/js-bindings/manual/jsb_cocos2dx_extension_manual.cpp [hunk 4]
```diff
@@ -126,6 +126,83 @@ static bool js_cocos2dx_extension_loadRemoteImage(se::State& s)
 }
 SE_BIND_FUNC(js_cocos2dx_extension_loadRemoteImage)
 
+static bool js_cocos2dx_extension_initRemoteImage(se::State& s)
```
Comment (jareguo) [inline cocos/scripting/js-bindings/manual/jsb_cocos2dx_extension_manual.cpp]: add a new binding interface like js_cocos2dx_extension_loadRemoteImage

File: cocos/renderer/CCTextureCache.h [hunk 5]
```diff
@@ -115,6 +115,14 @@ class CC_DLL TextureCache : public Ref
     */
     virtual void addImageAsync(const std::string &filepath, const std::function<void(Texture2D*)>& callback);
 
+    /** Init a given Texture2D object with a file image.
+    * It is much the same as addImageAsync, except an existing texture is provided.
+     @param texture An existing texture.
+     @param filepath A null terminated string.
+     @param callback A callback function would be invoked after the image is loaded.
+    */
+    virtual void initImageAsync(Texture2D* texture, const std::string& filepath, const std::function<void(Texture2D*)>& callback);
```
Comment (dumganhar) [inline cocos/renderer/CCTextureCache.h]: This interface is strange. 
`TextureCache::addImageAsync` is used for parsing image file to texture asynchronously, but if you have gotten a texture and you don't want to add it to `TextureCache`, why to add this interface to `TextureCache`? 
Is it ok to just use your `Texture2D*` instance directly?

Comment (jareguo) [reply cocos/renderer/CCTextureCache.h]: Just because I want to use the async queue in textureCache.

Comment (dumganhar) [reply cocos/renderer/CCTextureCache.h]: How about using `Scheduler::performFunctionInCocosThread` or `Scheduler::schedule` with repeat once to achieve this functionality?

**Comment by jareguo** [comment]
done, added a new binding js_cocos2dx_extension_initTextureAsync, ThreadPool also moved.

File: cocos/scripting/js-bindings/manual/jsb_cocos2dx_extension_manual.cpp [hunk 6]
```diff
@@ -132,6 +134,180 @@ static bool js_cocos2dx_extension_loadRemoteImage(se::State& s)
...
+
+        args[0].setBoolean(success);
+        func.toObject()->call(args, nullptr);
+    };
+
+    getThreadPool()->pushTask([=](int tid){
+        bool success = false;
+
+        Data data = FileUtils::getInstance()->getDataFromFile(url);
+        if (data.isNull() == false)
```
Comment (dumganhar) [inline cocos/scripting/js-bindings/manual/jsb_cocos2dx_extension_manual.cpp]: --> if (!data.isNull())

File: cocos/scripting/js-bindings/manual/jsb_cocos2dx_extension_manual.cpp [hunk 7]
```diff
@@ -132,6 +134,180 @@ static bool js_cocos2dx_extension_loadRemoteImage(se::State& s)
...
+        args.resize(1);
+
+        args[0].setBoolean(success);
+        func.toObject()->call(args, nullptr);
+    };
+
+    getThreadPool()->pushTask([=](int tid){
+        bool success = false;
+
+        Data data = FileUtils::getInstance()->getDataFromFile(url);
```
Comment (dumganhar) [inline cocos/scripting/js-bindings/manual/jsb_cocos2dx_extension_manual.cpp]: getDataFromFile isn't thread safe.
you could try to get the full path of url before pushTask, and use `lambda` to capture the std::string of full path.

**Pull request merged**